frontend/src/lib/ReportNode.ts

```diff
@@ -320,37 +311,54 @@ export class ReportRoot {
       mainNodes = []
     }
 
-    return new ReportRoot(
-      new BlockNode(
-        mainNodes,
-        BlockProto.create({ allowEmpty: true }),
-        NO_REPORT_ID
-      ),
-      BlockNode.empty()
+    const main = new BlockNode(
+      mainNodes,
+      new BlockProto({ allowEmpty: true }),
+      NO_REPORT_ID
     )
+
+    const sidebar = new BlockNode(
+      [],
+      new BlockProto({ allowEmpty: true }),
+      NO_REPORT_ID
+    )
+
+    return new ReportRoot(new BlockNode([main, sidebar]))
+  }
+
+  public constructor(root: BlockNode) {
+    this.root = root
+
+    // Verify that our root node has exactly 2 children: a 'main' block and
```

Comment:
I _think_ what you're looking for is what already exists. This "root" node has two children, `main` and `sidebar`, which are logically siblings (they have two separate top-level deltagenerators). There's no "root" DeltaGenerator or anything - the root node exists here only as an implementation detail to `ReportRoot`.

If we add new top-level DeltaGenerators, they'll just be additional siblings to `main` and `sidebar`, and all live under root.

Reply:
Yeah -- the reason I'm bringing this up is because in our Python backend, our DeltaGenerators.py classes are _also_ structured as trees (each DG has a "parent" pointer).[1] But subtly, the tree structure in Python is a different than the structure here in JS; in Python the [Sidebar DG is a child of the Main DG](https://github.com/streamlit/streamlit/blob/0b8505befef010abe3c5f9cf095f1a4af872ecbe/lib/streamlit/__init__.py#L108).

Again, no action item here, but this could be a subtle source of confusion in the future, and I want to think through how we could unify the two structures in a way that makes sense. Off the top of my head, the Python could copy the JS structure with a dummy Root class, or vice-versa (JS could copy Python)

[1] Originally this was so that the "with" DG could be tracked in the tree's root, which was the Main DG. Since then we moved that to ReportContext to fix the threading issue. At the moment we're just using it to check that columns are not nested.

Reply:
Ah cool. In my mind, having an (element-free) root that just contains the top-level root containers makes the most sense. I'd advocate for making that change at some point on the Python side.